proposals/0000-linear-types.rst

```diff
@@ -903,49 +903,86 @@ cope with affine functions. Therefore the choice between these three
-It would seem an arbitrary limitation for little gain to prevent them
-from using the linear types that they need. Plus even in ``RIO`` code,
-where exceptions do matter, linear types are useful: they allow prompt
-deallocation as argued in the section on Exceptions_, it can be much
-harder to reason on the lifetime of resources with explicit scopes
-like with ``bracket`` (see the `inline-java use-case
-<http://www.tweag.io/posts/2017-11-29-linear-jvm.html>`_ for an
-example where scopes have proved to be unsatisfactory).
-
-There is, nonetheless, value to affine types. There are some
-applications where affine types are enough to enforce invariants (such
-as in-place mutation of garbage-collected structure, like mutable
-arrays). And they can presumably benefit from the additional
-flexibility. For instance, ``catch`` can get a more fine-grained type
-(writing ``'A`` for the affine multiplicity):
+Exceptions can only be caught in the ``IO`` monad, yet this API is
+pure. So exceptions are not a concern in the design of this API. The
+properties this API wants to enforce hold even with linear types and
+even in the face of exceptions being thrown (in a pure or impure
+context) and caught (in an impure context). No linear types means this
+API would need to use CPS, if that works at all to enforce the same
+properties.
+
+Another example is `language interop
+<http://www.tweag.io/posts/2017-11-29-linear-jvm.html>`_ by
+@facundominguez and @mboes. In this example, Haskell users create GC
+roots for every object in the JVM's heap that they want to reference
+directly. These GC roots must be released as soon as the reference is
+no longer useful, otherwise memory pressure on the JVM will increase
+and throughput will suffer because of more frequent GC passes. Worse,
+deleting roots too late can sometimes prevent any memory from being
+reclaimed at all, leading to out-of-memory conditions. Introducing
+a ``bracket``-like ``withJvmScope`` action is one way to ensure all
+roots do get deleted eventually (at scope exit), but in practice, in
+complex dual-language projects, introducing neither too fine-grained
+or too coarse-grained scopes has proven very difficult. Furthermore,
+``bracket``-like constructs break tail-recursion. Linear types enable
+working with a single global resource scope, while still guaranteeing
+eventual deletion of roots, in any order. Affine types do not. At any
+rate, not in direct-style.
+
+Now, in this latter example, exceptions do impose both an
+implementation cost and a design cost. The implementation cost arises
+because we want a stronger guarantee: we want to know that all GC
+roots are always freed exactly once, so we must register each GC root
+to free them if an exception is thrown. A free-at-most-once guarantee
+wouldn't require this. The design cost is that ``catch`` requires
+a weaker type than desirable, as discussed above, limiting its power.
```

Comment:
I'm not sure I follow the point here.

First, on a purely technical stand-point: if I understand correctly we don't do any registering, Java does (with `pushLocalFrame`, `popLocalFrame` module the actual names).

Second, even if we had to do the registration thing, affine types would no alleviate that cost (if anything, it would make it more acute). It's just that it would apply both in case of normal exit and of exception. The point about `catch` still stands though.

Reply:
In our particular case it's the JNI that bears this implementation cost. So I could say there's actually no cost here, but in general that's not true, so I let it slide. Good point about affine types needing same in practice, even if it's not needed to comply with the protocol the types demand. I'll have to reword.